File: src/types/rag.ts
// ============= RAG 知识库 / 导学终端类型 =============

/** RAG 检索来源 */
export interface RAGSource {
  document: string
  page?: number
  section?: string
  content: string
  score: number // 相似度 0-1
  ref: number // 引用编号
}

/** RAG Token 用量 */
export interface RAGTokenUsage {
  prompt_tokens: number
  completion_tokens: number
  total_tokens: number
}

/** RAG 查询请求 */
export interface RAGQueryRequest {
  question: string
  top_k?: number
  temperature?: number
  max_tokens?: number
  student_id?: string
  subject?: string
  /** true = 跳过知识库检索直连大模型；默认 false（走 RAG 检索增强） */
  skip_retrieval?: boolean
  /** true = 快速模式（低温度、短回复、短超时） */
  fast_mode?: boolean
  /** true = 测绘模式（附加学习状态评估 JSON） */
  cehui_mode?: boolean
  /** true = 走 SSE 流式返回（仅 ragQueryStream 使用） */
  stream?: boolean
  /** 会话 ID（可选）：用于把问答存入会话历史，支撑画像提炼（建议 10） */
  sessionId?: string
}

/** RAG 查询响应 */
export interface RAGQueryResponse {
  answer: string
  sources: RAGSource[]
  confidence: number
  retrieval_count: number
  model: string
  token_usage?: RAGTokenUsage
  query_id: string
  /** 测绘模式下的学习评估数据 */
  cehui?: {
    mastery_estimates?: Array<{ kp_name: string; level: number }>
    cognitive_load?: number
    learning_intent?: string
    needs_optimization?: boolean
  }
}

/** RAG 知识库统计 */
export interface RAGStats {
  collection_name: string
  total_documents: number
  total_chunks: number
  indexed_files: string[]
}

/** RAG 索引请求 */
export interface RAGIndexRequest {
  directory: string
  recursive?: boolean
}

/** RAG 索引响应 */
export interface RAGIndexResponse {
  total_chunks: number
  indexed_files: string[]
  errors: string[]
}

// ============= 对话前端类型 =============

/** 对话角色 */
export type ChatRole = 'user' | 'assistant' | 'system'

/** 学科选项 */
export interface SubjectOption {
  label: string
  value: string
  description?: string
}

/** 单条对话消息 */
export interface ChatMessage {
  id: string
  role: ChatRole
  content: string
  sources?: RAGSource[]
  confidence?: number
  tokenUsage?: RAGTokenUsage
  timestamp: number
  isStreaming?: boolean // 是否正在流式输出
  queryId?: string
  /** 建议 9：助手消息是否含可复制使用素材（代码块/提纲），触发反思框 */
  hasReusableMaterial?: boolean
  /** 建议 9：反思框状态机: locked -> reflecting -> unlocked */
  reflectState?: 'locked' | 'reflecting' | 'unlocked'
  /** 建议 9：是否已勾选"我已读懂" */
  reflectAcknowledged?: boolean
  /** 建议 9：最近一次反思判定结果 */
  reflectResult?: { understood: boolean; feedback: string; followUp: string } | null
}

/** 快捷提问项 */
export interface QuickQuestion {
  id: string
  text: string
  icon?: string
}

/** 对话导出格式 */
export interface ChatExportData {
  exportTime: string
  subject: string
  messages: {
    role: ChatRole
    content: string
    sources?: {
      document: string
      page?: number
      content: string
    }[]
    timestamp: number
  }[]
}

File: backend/app/schemas/rag.py
"""RAG 知识库相关 Schema"""

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class RAGQueryRequest(BaseModel):
    """RAG 问答请求"""

    question: str = Field(..., min_length=1, max_length=2000, description="用户问题")
    top_k: int = Field(default=5, ge=1, le=20, description="检索返回数量")
    temperature: float = Field(default=0.5, ge=0.0, le=2.0, description="LLM 温度")
    max_tokens: int = Field(default=1024, ge=64, le=4096, description="最大生成 token 数")
    student_id: Optional[str] = Field(default=None, description="学生 ID（可选）")
    subject: Optional[str] = Field(default=None, description="学科过滤（可选）")
    skip_retrieval: bool = Field(default=False, description="跳过知识库检索，直接使用大模型回答")
    stream: bool = Field(
        default=False,
        description="是否以 SSE (text/event-stream) 流式返回。默认 False 保持整包 JSON 响应",
    )


class RAGSource(BaseModel):
    """问答来源"""

    document: str = Field(..., description="文档名称")
    page: str = Field(default="", description="页码")
    section: str = Field(default="", description="章节")
    content: str = Field(default="", description="引用内容摘要")
    score: float = Field(default=0.0, description="相似度得分")
    ref: int = Field(default=0, description="引用编号")


class RAGTokenUsage(BaseModel):
    """Token 用量"""

    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0


class RAGQueryResponse(BaseModel):
    """RAG 问答响应"""

    answer: str = Field(..., description="AI 生成的答案")
    sources: List[RAGSource] = Field(default_factory=list, description="引用来源列表")
    confidence: float = Field(default=0.0, ge=0.0, le=1.0, description="置信度")
    retrieval_count: int = Field(default=0, description="检索到的文档块数量")
    model: str = Field(default="", description="使用的模型")
    token_usage: Optional[RAGTokenUsage] = Field(default=None, description="Token 用量")
    query_id: str = Field(default="", description="查询追踪 ID")


class RAGIndexRequest(BaseModel):
    """索引请求"""

    directory: str = Field(..., description="要索引的文档目录路径")
    recursive: bool = Field(default=True, description="是否递归子目录")
    clear_existing: bool = Field(default=False, description="是否先清空已有索引")


class RAGIndexResponse(BaseModel):
    """索引响应"""

    message: str = Field(..., description="结果消息")
    chunks_indexed: int = Field(default=0, description="索引的文档块数量")
    directory: str = Field(default="", description="索引的目录")


class RAGStatsResponse(BaseModel):
    """知识库统计"""

    status: str = Field(default="not_initialized", description="状态")
    documents: int = Field(default=0, description="文档数量")
    name: str = Field(default="", description="集合名称")
    directory: str = Field(default="", description="持久化目录")
    similarity_threshold: float = Field(default=0.5, description="相似度阈值")
    chunk_config: Dict[str, Any] = Field(default_factory=dict, description="分块配置")
